diesel/src/main.rs

```diff
@@ -88,46 +88,44 @@ fn index_add(req: &HttpRequest<AppState>) -> Box<Future<Item = HttpResponse, Err
         })
         // `Future::and_then` can be used to merge an asynchronous workflow with a
         // synchronous workflow
-        .and_then(|body| {
+        //
+        // Douman NOTE:
+        // The return value in this closure helps, to clarify result for compiler
+        // as otheriwse it cannot understand it
+        .and_then(move |body| -> Box<Future<Item = HttpResponse, Error = Error>> {
             // body is loaded, now we can deserialize serde-json
             let r_obj = serde_json::from_slice::<MyUser>(&body);
 
             // Send to the db for create
             match r_obj {
                 Ok(obj) => {
-                req.state()
-                    .db
-                    .send(CreateUser {
-                        name: obj.name,
-                    })
-                    .from_err()
-                    .and_then(|res| match res {
-                        Ok(user) => Ok(HttpResponse::Ok().json(user)),
-                        Err(_) => Ok(HttpResponse::InternalServerError().into()),
-                    })
-                }
-                Err(_) => {
-                    Err(error::ErrorBadRequest("Json Decode Failed"))
+                let res = state.db.send(CreateUser { name: obj.name, })
+                               .from_err()
+                               .and_then(|res| match res {
+                                   Ok(user) => Ok(HttpResponse::Ok().json(user)),
+                                   Err(_) => Ok(HttpResponse::InternalServerError().into()),
+                               });
+
+                Box::new(res)
                 }
+                Err(_) => Box::new(future::err(error::ErrorBadRequest("Json Decode Failed")))
             }
         })
-        .responder()
 }
 
-fn add2((item, req): (Json<MyUser>, HttpRequest<AppState>)) -> Box<Future<Item = HttpResponse, Error = Error>> {
-    Box::new(req.state()
-        .db
-        .send(CreateUser {
-            // Is it possible to do with without cloning? Apparently this is "borrowed".
-            // A risk is that when this is sent as a message, the lifetimes will
-            // be very complex to handle, and may not be reasonable ...
-            name: item.name.clone(),
-        })
-        .from_err()
-        .and_then(|res| match res {
-            Ok(user) => Ok(HttpResponse::Ok().json(user)),
-            Err(_) => Ok(HttpResponse::InternalServerError().into()),
-        }))
+fn add2((item, state): (Json<MyUser>, State<AppState>)) -> impl Future<Item = HttpResponse, Error = Error> {
+    state.db
+         .send(CreateUser {
+             // Is it possible to do with without cloning? Apparently this is "borrowed".
+             // A risk is that when this is sent as a message, the lifetimes will
+             // be very complex to handle, and may not be reasonable ...
+             name: item.name.clone(),
```

Comment:
It is borrowed due to `Deref`, use `Json::into_innter` to get `MyUser` struct and then move out of it